CRM/Core/Payment/PayPalIPN.php

```diff
@@ -321,9 +321,8 @@ public function main() {
       $ids['onbehalf_dupe_alert'] = self::retrieve('onBehalfDupeAlert', 'Integer', 'GET', FALSE);
     }
 
-    $paymentProcessorID = CRM_Core_DAO::getFieldValue('CRM_Financial_DAO_PaymentProcessorType',
-      'PayPal_Standard', 'id', 'name'
-    );
+    // CRM-18302: get payment processor ID passed from CRM_Core_Payment_PayPalImpl::doTransferCheckout()
+    $paymentProcessorID = self::retrieve('paymentProcessorID', 'Integer', 'GET', TRUE);
```

Comment:
You need to make sure that $paymentProcessorID is always set. Like following pseudo 
 IF 
$paymentProcessorID = self::retrieve('paymentProcessorID', 'Integer', 'GET', TRUE);
  ELSE 
$paymentProcessorID = CRM_Core_DAO::getFieldValue('CRM_Financial_DAO_PaymentProcessorType', 'PayPal_Standard', 'id', 'name');
ENDIF

Also since https://github.com/civicrm/civicrm-core/pull/8012 PR is merged your current PR is in conflict, so please resolve the conflicts on rebase with changes stated above and update the PR. Rest look fine to me.